// envdist/dist/src/services/parser/env.parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class EnvParser {
    constructor() {
        this.countEnvGroups = 3;
        this.countDistGroups = 4;
        this.countDistGroupsWithDefault = 6;
        this.distRegex = /^([A-Za-z_0-9]{1,255})[\s\t]*=[\s\t]*(\<(.+)\>)([\s\t]*\#[\s\t]*(.+))?$/m;
        this.envRegex = /^([A-Za-z_0-9]{1,255})[\s\t]*=[\s\t]*([^\s].*)$/m;
    }
    parseDist(line) {
        const match = line.match(this.distRegex);
        if (!match || (match.length !== this.countDistGroups && match.length !== this.countDistGroupsWithDefault)) {
            return null;
        }
        return {
            name: match[1],
            description: match[3],
            default: match[5],
        };
    }
    parseEnv(line) {
        const match = line.match(this.envRegex);
        if (!match || (match.length !== this.countEnvGroups)) {
            return null;
        }
        return {
            name: match[1],
            value: match[2],
        };
    }
}
exports.EnvParser = EnvParser;
